ql/src/java/org/apache/hadoop/hive/ql/Compiler.java

```diff
@@ -72,19 +65,52 @@
   private static final Logger LOG = LoggerFactory.getLogger(CLASS_NAME);
   private static final LogHelper CONSOLE = new LogHelper(LOG);
 
-  private final Context context;
-  private final DriverContext driverContext;
+  protected final Context context;
+  protected final DriverContext driverContext;
   private final DriverState driverState;
   private final PerfLogger perfLogger = SessionState.getPerfLogger();
 
   private ASTNode tree;
 
+  /*
+   * Compiler needs to know if given statement is PREPARE/EXECUTE or otherwise
+   * to orchestrate the planning.
+   * e.g. if plan should be stored or if parameter binding should be done.
+   */
+  enum StatementType{
+    PREPARE,
+    EXECUTE,
+    REGULAR
+  };
+
   public Compiler(Context context, DriverContext driverContext, DriverState driverState) {
     this.context = context;
     this.driverContext = driverContext;
     this.driverState = driverState;
   }
 
+  /*
+   * Give string command determine the type of statement
+   */
+  private StatementType getStatementType(final ASTNode tree) {
```

Comment:
I was hoping this orchestration could be moved to the different analyzers for `PREPARE` and `EXECUTE`, e.g., similar to `ExplainSemanticAnalyzer` or others. It may make the code easier to understand instead of creating different workflows in the compiler depending on the statement type (I think `Compiler` is mostly agnostic from statement type and would probably be good to try to keep it that way).